test/infrastructure/docker/e2e/docker_upgrade_test.go

```diff
@@ -0,0 +1,265 @@
+// +build e2e
+
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/test/framework"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Docker Upgrade", func() {
+	var (
+		replicas     = 3
+		namespace    = "default"
+		clusterGen   = newClusterGenerator("upgrade")
+		mgmtClient   ctrlclient.Client
+		cluster      *clusterv1.Cluster
+		controlPlane *controlplanev1.KubeadmControlPlane
+	)
+	SetDefaultEventuallyTimeout(10 * time.Minute)
+	SetDefaultEventuallyPollingInterval(10 * time.Second)
+
+	BeforeEach(func() {
```

Comment:
Since this part is mostly same with Basic test, instead of separating them to multiple files, why not keep them as is so that beforeach and after each code can be reused?

Reply:
I made it separate because the original test had failure domains configured and I didn't want to couple the failure domain assertions with the other upgrade tests. 
TBH I'm currently unfamiliar with failure domains and as I was trying to write these tests up I wanted to do the upgrade against a cluster that was "as vanilla as I could get".

Reply:
Having failure domains should not cause issue and checking that logic is important too. I think this is fine for this PR, but we should consider combining these files for not maintaining before/after suites in two places and also AFAIK to run it in parallel, we will need to have them together.

Reply:
We have a single before/after suites to create the management cluster. 
The Before/After each is responsible for creating workload clusters (with different identifiers) and because of that isolation I believe they can run in parallel.

I also think that keeping these tests separate would make them more readable and provide a better understanding of what the expectations are. 
Initially, as someone who did not have context, when I came to these tests I didn't know if the upgrades somehow relied on the failure domain configuration because the original full upgrade test was running against a cluster with failure domains configured.

That being said, if others feel that it would be better to couple these tests together I don't mind doing that in a follow up PR.

Reply:
+1 to separating them but mitigating the speed hit with running them in parallel.

Reply:
I can create a separate issue to track the work of making these tests run in parallel.

Reply:
If we're duplicating code, I'd suggest to make the similar code in a different function and call it from multiple places

Reply:
Issue for running CAPD e2e in parallel https://github.com/kubernetes-sigs/cluster-api/issues/2795